File: arrays/summaryRanges.go
package main

import (
	"fmt"
	"strings"
)
/*

link leetcode problem -> https://leetcode.com/problems/summary-ranges/

time complexity  -> O(n)
space complexity -> O(n)

*/
func summaryRanges(nums []int) []string {
	if len(nums) == 0 {
		return []string{}
	}
	var stringResult []string
	var sequence strings.Builder
	startSequence := nums[0]
	endSequence := nums[0]

	i:= 0
	j:= 1

	for j < len(nums) {
		if nums[i] == nums[j] - 1{
			endSequence = nums[j]
		}else{
			sequence.WriteString(fmt.Sprintf("%d->%d", startSequence, endSequence))
			
			if startSequence == endSequence{
				sequence.Reset()
				sequence.WriteString(fmt.Sprintf("%d",endSequence))
				
			} 
			stringResult = append(stringResult,sequence.String())
			sequence.Reset()
			startSequence = nums[j]
			endSequence = nums[j]
		}
		i++
		j++
	}

	sequence.WriteString(fmt.Sprintf("%d->%d", startSequence, endSequence))
	if startSequence == endSequence{
		sequence.Reset()
		sequence.WriteString(fmt.Sprintf("%d",endSequence))
		
	} 
	stringResult = append(stringResult,sequence.String())
	sequence.Reset()

	return stringResult

}

File: arrays_and_strings/summary_ranges.py
#Link to problem: https://leetcode.com/problems/summary-ranges/description/
#Time complexity: O(n)
#Space complexity: O(n)
                       
class Solution:
    def summary_ranges(self,nums:list[str]) -> list[str]:
        if len(nums) == 0:
            return []

        begin_sequence = nums[0]
        end_sequence = nums[0]
        sequence = ""
        ranges = []
        
        for num in nums:
            if num > end_sequence + 1:
                sequence = f"{begin_sequence}->{end_sequence}" if begin_sequence != end_sequence else f"{begin_sequence}"
                ranges.append(sequence)
                begin_sequence = num
            
            end_sequence = num
        
        sequence = f"{begin_sequence}->{end_sequence}" if begin_sequence != end_sequence else f"{begin_sequence}"
        ranges.append(sequence)
        
        return ranges
                                        
solution = Solution()
print(solution.summary_ranges([0,1,2,4,5,7]))
print(solution.summary_ranges([0,2,4,6,8,10,12,14]))
print(solution.summary_ranges([0]))
print(solution.summary_ranges([]))
